packages/cozy-realtime/src/index.js

```diff
@@ -99,228 +120,233 @@ const configTypes = {
-          console.warn(`Reconnecting ... ${numRetries} tries left.`)
-          setTimeout(async () => {
-            try {
-              socket = await connectWebSocket(
-                config,
-                onSocketMessage,
-                onSocketClose,
-                --numRetries,
-                retryDelay + 1000
-              )
-            } catch (error) {
-              // eslint-disable-next-line no-console
-              console.error(
-                `Unable to reconnect to realtime. Error: ${error.message}`
-              )
-            }
-          }, retryDelay)
-        }
-      }
+    if (listenerKey === payload.type) return
+
+    // doctype listner call
+    if (listeners.has(payload.type) && listeners.get(payload.type)[eventType]) {
+      listeners.get(payload.type)[eventType].forEach(listener => {
+        listener(payload.doc)
+      })
     }
+  }
 
-    try {
-      socket = await connectWebSocket(
-        config,
-        onSocketMessage,
-        onSocketClose,
-        NUM_RETRIES,
-        RETRY_BASE_DELAY
+  const onSocketClose = (event, numRetries, retryDelay) => {
+    if (!event.wasClean) {
+      console.warn(
+        `WebSocket closed unexpectedly with code ${event.code} and ${
+          event.reason ? `reason: '${event.reason}'` : 'no reason'
+        }.`
       )
-    } catch (error) {
-      reject(error)
-    }
 
-    cozySocket = {
-      subscribe: (doctype, event, listener) => {
-        if (typeof listener !== 'function')
-          throw new Error('Realtime event listener must be a function')
-
-        if (!listeners[doctype]) {
-          listeners[doctype] = {}
-          subscribeWhenReady(doctype, socket)
-        }
-
-        listeners[doctype][event] = (listeners[doctype][event] || []).concat([
-          listener
-        ])
-      },
-      unsubscribe: (doctype, event, listener) => {
-        if (
-          listeners[doctype] &&
-          listeners[doctype][event] &&
-          listeners[doctype][event].includes(listener)
-        ) {
-          listeners[doctype][event] = listeners[doctype][event].filter(
-            l => l !== listener
-          )
-        }
+      if (numRetries) {
+        console.warn(`Reconnecting ... ${numRetries} tries left.`)
+        setTimeout(() => {
+          try {
+            socket = connectWebSocket(
+              config,
+              onSocketMessage,
+              onSocketClose,
+              --numRetries,
+              retryDelay + 1000,
+              true
+            )
+          } catch (error) {
+            console.error(
+              `Unable to reconnect to realtime. Error: ${error.message}`
+            )
+          }
+        }, retryDelay)
       }
     }
+  }
 
-    resolve(cozySocket)
-  })
-}
+  try {
+    socket = connectWebSocket(
+      config,
+      onSocketMessage,
+      onSocketClose,
+      NUM_RETRIES,
+      RETRY_BASE_DELAY
+    )
+  } catch (error) {
+    throw error
```

Comment:
Same remark as the one above made by @nono, we could avoid the try catch here if it is just used to throw the same error.